flink-yarn-tests/src/test/java/org/apache/flink/yarn/YarnTestBase.java

```diff
@@ -573,6 +576,23 @@ public static int getRunningContainers() {
 		return count;
 	}
 
+	void waitUntilApplicationFinished(ApplicationId applicationId, Duration timeout) throws Exception {
+		Deadline deadline = Deadline.now().plus(timeout);
+		while (true) {
+			YarnApplicationState state = yarnClient.getApplicationReport(applicationId).getYarnApplicationState();
+			if (state == YarnApplicationState.FINISHED) {
+				break;
+			} else if (state == YarnApplicationState.FAILED || state == YarnApplicationState.KILLED) {
+				Assert.fail("Application became FAILED or KILLED while expecting FINISHED");
+			} else {
+				sleep(sleepIntervalInMS);
+			}
+			if (deadline.isOverdue()) {
+				Assert.fail("Application didn't finish before timeout");
```

Comment:
I think we should move the app killing from the finally block of `YARNITCase. waitUntilApplicationFinished` to this timeout case, right before `Assert.fail("Application didn't finish before timeout");`. As we need killing only if the app hangs but we do not need if it finished successfully.

Reply:
I put the `waitUntilApplicationFinished` into YarnITCase as it is not so universal for other case now.

Reply:
agreed